### ISSUE TITLE
Please support Rust-free `librsvg`

### ISSUE THREAD

**Issue by barracuda156** [Beep6581/RawTherapee#7060]
Please support Rust-free `librsvg`, since Rust is broken on a number of platforms with unclear timeline when it gets fixed if ever.
And unfortunately, Gnome folks require Rust for the current `librsvg` versions.

**Comment by Lawrence37** [comment]
librsvg 2.52 or higher is required for hi-dpi, so there may not be a good solution.

**Comment by barracuda156** [comment]
> librsvg 2.52 or higher is required for hi-dpi, so there may not be a good solution.

@Lawrence37 You mean for the output for Retina displays? It is perfectly fine not to have it on systems without Rust (which are likely to be either non-mainstream or not too powerful).

Rather than losing the functionality, it is much preferable to have it reasonably limited. I.e., retain the codepath for earlier `librsvg` when newer one is not detected.

P. S. Some people may also have security concerns with Rust, since it can only be compiled with itself presently.

**Comment by Lawrence37** [comment]
Hi-dpi required some significant changes, so it may not be feasible to make it optional. That's what I meant by "no good solution". @Pandagrapher knows more about it than me.

**Comment by barracuda156** [comment]
@Lawrence37 Oh, that would kill RawTherapee for all systems without Rust. That would be really bad…

**Comment by RJVB** [comment]
> Hi-dpi required some significant changes, so it may not be feasible to make it optional. That's what I meant by "no good solution". @Pandagrapher knows more about it than me.

I don't plan to get involved here but I think there must be a possibility to pipe the output of older, normal-res librsvg routines into the hi-res code path, via a scale factor?

Or make the entire support for SVG optional so at least all the other features continue to work? (SVG isn't exactly a photo format in my book...)

**Comment by Lawrence37** [comment]
That requires maintaining the old code along with the new code, which is exactly the option that I consider infeasible. RawTherapee doesn't support editing SVG images. We only use SVG for the icons in the GUI.

**Comment by barracuda156** [comment]
@Lawrence37 Is it really worth it for the sake of icons in the gui?

If it can be done in a modular way, that will also solve the problem. I don’t care about icons in the gui, can I have conversion engine etc. updated while staying with existing gui?

**Comment by Lawrence37** [comment]
Are you saying HiDPI support is not worth it?

**Comment by barracuda156** [comment]
@Lawrence37 AFAIK, pretty much all apps using GTK in MacPorts work at the moment on systems without Retina support and without Rust in the same versions which work on the latest macOS. I did not look into how it is done, to be honest. GTK3 has some glitches with GUI on arm64, GTK4 works fine there (it requires manual env setting on legacy macOS, but no special codepath, AFAIK).

I am not the authority to judge what is worth, of course. From my side it looks that this should be doable in some way with minimal effort without a burden of maintaining a dedicated codepath, since this seems to be the case generally.

Regardless of that though, a modular option will be a solution and perhaps a general improvement. As long as core components can be updated separately from GUI, we are good.

**Comment by Benitoite** [comment]
@Hombre57 authored the switch to svg icons for hidpi support.  At the time, five years ago, there were people requesting hidpi support, and no one requesting macosx/ppc support, so...
I have to believe there is a way to get rust on ppc macosx.  Have you looked at any solutions like https://github.com/rust-lang/rustc_codegen_gcc

**Comment by Lawrence37** [comment]
I will let @Pandagrapher explain why RawTherapee needs the minimum version that it asks for, since I do not know the specifics.

A clean separation of the GUI is not possible without a complete redesign due to the tight coupling between the tools' UI and parameters.

**Comment by RJVB** [comment]
> That requires maintaining the old code along with the new code, which is exactly the option that I consider infeasible. 

Sorry, what is "that" here?

Did librsvg break or introduce new API when they introduced hi-res support? If not I don't see why you'd need to maintain any old code.
As a null approach you could just use whatever librsvg returns (some kind of bitmap imagine I presume), letting people who insist on using an old version put up with icons that are too small.

Actually, that might be exactly what you get when you tweak the build to use an old librsvg version?

> We only use SVG for the icons in the GUI.

Doesn't GTk have functions for loading and rendering icons that work as you'd expect regardless of screen resolution?! And that would accept PNG versions too (preferable anyway in my book: disk space is cheap, CPU cycles not).

**Comment by Benitoite** [comment]
> Sorry, what is "that" here?

Going back to png icons.

**Comment by Pandagrapher** [comment]
> Hi-dpi required some significant changes, so it may not be feasible to make it optional. That's what I meant by "no good solution". @Pandagrapher knows more about it than me.

We require `librsvg` 2.52 or higher for the `get_intrinsic_size_in_pixels` function (added from this version) that allows retrieving the SVG size in pixels. This function is required even for non-HiDPI screen (we just perform a scaling or not according to HiDPI setting).


> Doesn't GTk have functions for loading and rendering icons that work as you'd expect regardless of screen resolution?! And that would accept PNG versions too (preferable anyway in my book: disk space is cheap, CPU cycles not).

RT 5.10 now supports icon themes that means icons can be prodivided in PNG and/or SVG. Such as more and more Gtk app (e.g. Inkscape), we only provide SVG version of icon theme but you still can generate PNG ones for each pixel size.

**Comment by barracuda156** [comment]
> > Hi-dpi required some significant changes, so it may not be feasible to make it optional. That's what I meant by "no good solution". @Pandagrapher knows more about it than me.
> 
> 
> 
> We require `librsvg` 2.52 or higher for the `get_intrinsic_size_in_pixels` function (added from this version) that allows retrieving the SVG size in pixels. This function is required even for non-HiDPI screen (we just perform a scaling or not according to HiDPI setting).

@Pandagrapher Maybe it is possible to add a fallback just for this one function? `librsvg` worked somehow without it pre-2.52.

**Comment by Pandagrapher** [comment]
> @Pandagrapher Maybe it is possible to add a fallback just for this one function? `librsvg` worked somehow without it pre-2.52.

If all the icons in RT icon theme are converted to PNG for different sizes, basically `librsvg` is not required anymore by RT. However, as some other Gtk app (e.g. Inkscape), we made the choice to only provide SVG only icon theme. As a workaround, you can locally convert the icons to PNG and stub use of `RTScalable::loadSurfaceFromIcon` function in code

**Comment by RJVB** [comment]
>We require `librsvg` 2.52 or higher for the `get_intrinsic_size_in_pixels` function (added from this version) that allows retrieving the SVG size in pixels.

That does sound like something that one could home-roll even if it's at the cost of doing an actual rendering of the graphic. (But then again I don't see what "intrinsic size in pixels"
means for a vector drawing so I may misunderstand something completely here.)

>as more and more Gtk app (e.g. Inkscape), we only provide SVG version of icon theme

What I was hinting at of course is that Qt has had built-in support for SVG-based icons since just about forever O:^)

**Comment by Benitoite** [comment]
Intrinsic size is internal to the svg, see https://gnome.pages.gitlab.gnome.org/librsvg/Rsvg-2.0/method.Handle.get_intrinsic_size_in_pixels.html

**Comment by RJVB** [comment]
>Intrinsic size is internal to the svg

As I thought, an unhappy name, it's just the intended size in pixels based on the SVG's intended size in physical units and the rendering density.

I suppose it should be possible to write an equivalent function in C/++, or maybe even translate (de-oxidise?) the existing rusty one.

Are there any constraints on the kind of physical units you are expecting (or would expect) in SVG icons?

How do you handle the case where the function returns false, either because there is no physical size set in the SVG or because it's not an actual physical size but a relative one?
In other words, would a dummy

```
gboolean rsvg_handle_get_intrinsic_size_in_pixels (
  RsvgHandle* , gdouble* , gdouble* )
{
    return FALSE;
}
```

be a workable fallback?

**Comment by Pandagrapher** [comment]
It's still possible to add custom code, specific fallback, ... The question is "Should we?"

For the following reasons, IMO we shouldn't develop a fix:
- Main Linux distros use a `librsvg` library version greater than 2.52 on their last LTS versions (e.g. Debian Bookwork or Ubuntu Jammy)
- Gtk4 have planned to directly use `librsvg` to render SVG (in the PR, minimum version is 2.52 for `librsvg`). When RT will start migrating to Gtk4, use of `librsvg` rust version would be de facto mandatory
- At RawTherapee level, it's better not to use custom code or callback at maximum for maintenability, bug correction and evolutibility. Indeed, RawTherapee is developped by a small group: using libraries allows having parallel development teams "helping us" having the best software, while we are focused on the RawTherapee dedicated parts (GUI configuration and image treatment engine)
- Using rust for `librsvg` is a `librsvg` developers choice

**Comment by barracuda156** [comment]
> Gtk4 have planned to directly use librsvg to render SVG (in the PR, minimum version is 2.52 for librsvg). When RT will start migrating to Gtk4, use of librsvg rust version would be de facto mandatory

As of now at least, `gtk4` works fine with rust-free `librsvg`. Even if that changes in the future, rust-free `gtk4` will continue to work for some years to come. There is no reason for 2024 software to suddenly get broken overnight :) 

> Using rust for librsvg is a librsvg developers choice

But it is a poor choice, re which quite a bit of people are unhappy (not only legacy macOS users).

**Comment by RJVB** [comment]
On Friday August 09 2024 08:55:14 Sergey Fedorov wrote:

>But it is a poor choice, re which quite a bit of people are unhappy (not only legacy macOS users).

I tend to agree with that if not only given the clear and significant overhead, but I would also argue that the poorer choice was by the GTk developers who accepted the new language dependency. Then again that's not really surprising given the Gnome record of doing things the way they prefer regardless of what others think.

Evidently my suggestion of implementing a fallback function would/could work for GTk too; the most transparent approach would be to add that function to a forked, stainless librsvg (preventing any addition maintenance bundle for dependent software).

**Comment by barracuda156** [comment]
@Pandagrapher @RJVB @Lawrence37 @Benitoite Turned out there are, apparently, alternatives to `librsvg`, which were pointed out to in a similar issue in another repo: https://github.com/hyprwm/hyprcursor/issues/83

Will it be possible to support some of those instead (or even switch away to)?

**Comment by RJVB** [comment]
On Wednesday April 02 2025 06:08:08 Sergey Fedorov wrote:

>@RJVB >: 
>>Will it be possible to support some of those instead (or even switch away to)?

No idea but I'd guess that it should be possible technically, and that PRs are always welcome ;)

**Comment by barracuda156** [comment]
@Pandagrapher @Lawrence37 Okay, since now we are at the point that latest releases are broken, what is more sensible approach to have it fixed? Reverting by hand something like https://github.com/RawTherapee/RawTherapee/commit/89d2bdce5b930fb20ec5c2ad764962a0dd16fa3a is crazy. Rewriting `librsvg` back to a normal language is not feasible – apparently it is not even bits & pieces like in some Python packages, but literally the whole thing is in rust now.

There is no necessity to fix anything here in the master, of course, I am fine fixing it in my fork. What can be considered besides reverting 2–3k LoC? 

There is no real alternatives to `RawTherapee`, so it is rather desirable to have it working and not being forever stuck at 5.10.

**Comment by digitalcarp** [comment]
It seems [lunasvg](https://github.com/sammycage/lunasvg) is pretty easy to integrate. It's all C/C++ with only [plutovg](https://github.com/sammycage/plutovg) as a dependency. I could try adding it as another svg renderer option? I'd prefer waiting for the current CMake PR to be merged first before trying though.

**Comment by barracuda156** [comment]
@digitalcarp That would be fantastic!

**Comment by RJVB** [comment]
>barracuda156 left a comment (RawTherapee/RawTherapee#7060)
>
>@digitalcarp That would be fantastic!

@barracuda156 - no idea how extensive the librsvg API is but it might be useful to write a set of compatibility/glue functions to make lunasvg at least an API-level drop-in replacement for librsvg? ;)

**Comment by digitalcarp** [comment]
I've branched off the CMake upgrade branch and implemented the replacement. You can try below, but it's behind compared to latest dev branch.

Please try building with `git clone --single-branch -b lunasvg https://github.com/digitalcarp/RawTherapee.git`. You need to specify `-DSVG_BACKEND=lunasvg` when configuring CMake. Default is `-DSVG_BACKEND=librsvg`.

**Comment by barracuda156** [comment]
@digitalcarp Good news – the build with `lunasvg` works!

Bad news – the binary segfaults, _however_ I am sure it has nothing to do with `lunasvg` or your branch: it looks precisely like the old bug which was there back in 5.10, introduced by https://github.com/RawTherapee/RawTherapee/commit/8accebe4161c581c76dda4edee027b95ec109ea9 commit.
See: https://github.com/RawTherapee/RawTherapee/issues/7052

This is the backtrace now:

<img width="781" height="1011" alt="Image" src="https://github.com/user-attachments/assets/64daf979-fbc5-451c-87cd-e962822b21d5" />

I think it is the same issue. My patch did not apply cleanly, so I dropped it in a hope it was fixed. It wasn’t.

I will try reverting that commit again to confirm it is the same bug and do a quick partial rebuild. Hopefully it works, otherwise I will take a break until tomorrow, need some sleep.

**Comment by digitalcarp** [comment]
I think there is currently a PR to fix your segfault: #7451

**Comment by barracuda156** [comment]
@digitalcarp That would be awesome. I will confirm tomorrow once back to the testing machine.

In the meanwhile, could we accommodate external libfmt, plutosvg and lunasvg? Otherwise the build installs everything in common directories, which gonna break stuff in any package management system.
```
Error: Failed to activate rawtherapee-devel: The following ports have active files that conflict with rawtherapee-devel's:
plutovg @1.1.0_0
  /opt/local/include/plutovg/plutovg.h
  /opt/local/lib/cmake/plutovg/plutovgConfig.cmake
  /opt/local/lib/cmake/plutovg/plutovgConfigVersion.cmake
```
If for some reason external dependencies specifically for these libs are not desirable, then perhaps they could be placed inside rawtherapee plugins directory (or something similar to the same effect).

**Comment by digitalcarp** [comment]
Try again. I made `fmt` not have a install target and made `lunasvg` be found via `PkgConfig`.

**Comment by barracuda156** [comment]
@digitalcarp Yes, will be tested today. On priority list.

**Comment by barracuda156** [comment]
> I think there is currently a PR to fix your segfault: [#7451](https://github.com/RawTherapee/RawTherapee/pull/7451)

@digitalcarp So do I need just to add 5 commits from https://github.com/digitalcarp/RawTherapee/commits/meyers on top of `lunasvg` branch?

Update. This https://github.com/digitalcarp/RawTherapee/commit/1e2557f547c6a9b6ac8d44c622c81c5457b2abfc commit does not apply on top of `lunasvg`. It is about 900 LoC…
What will be a more reasonable way to test `lunasvg`-based build? Or should I wait until it gets merged into the upstream branch?
As of now, there are three options, but neither is trivial:
1. Rebase `lunasvg` on top of `meyers`.
2. Cherry-pick from `meyers` to `lunasvg`, adjusting for all differences between `lunasvg` and `dev`.
3. Revert https://github.com/RawTherapee/RawTherapee/commit/8accebe4161c581c76dda4edee027b95ec109ea9 from `lunasvg` (does not work easily either, need to factor in code changes since 5.10).

**Comment by barracuda156** [comment]
@digitalcarp Once you have time, please give me some advice here. I wanna test this, but I need to combine both lunasvg part and a fix for segfault, which are currently in diverged branches.

Otherwise lunasvg part is building fine. I just can’t confirm if it works.

**Comment by digitalcarp** [comment]
I merged the two branches. They are based off of latest dev. It should at least compile + open without an immediate segfault.

`git clone --single-branch -b lunasvg-rebase https://github.com/digitalcarp/RawTherapee.git`

**Comment by barracuda156** [comment]
Awesome, I will try this today then.

**Comment by barracuda156** [comment]
@digitalcarp It works now!

<img width="1920" height="1004" alt="Image" src="https://github.com/user-attachments/assets/1ae2499a-e3bb-4afa-8716-af258397787a" />

**Comment by digitalcarp** [comment]
Great! Hopefully in a month or so all these changes will be merged into dev.

**Comment by barracuda156** [comment]
Thank you!
